### PR TITLE
Add MetaTensor support to sharding APIs.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78325

For very large models, it is useful to create an nn.Module on
device=meta and then shard it as desired.

As a result, adding support to our sharding APIs to support meta tensor.

Differential Revision: [D36685070](https://our.internmc.facebook.com/intern/diff/D36685070/)

#Closes: https://github.com/pytorch/pytorch/issues/65569

### CODE REVIEW

**Comment by janeyx99** [comment]
> @janeyx99 I was wondering why multigpu tests don't run by default on all distributed PRs?

The multigpu tests are a part of a trunk-only workflow as they use large linux16x machines. Currently the default is to run pull workflows on pull and to only trigger trunk when ciflow/trunk is specified. 

If we wanted multigpu tests to run for every workflow, we can discuss moving it to be based on the 11.3 build, which is run on every PR. @seemethere any thoughts about potentially increasing our need for linux16x machines?

**Comment by pritamdamania87** [comment]
> The multigpu tests are a part of a trunk-only workflow as they use large linux16x machines. Currently the default is to run pull workflows on pull and to only trigger trunk when ciflow/trunk is specified.

> If we wanted multigpu tests to run for every workflow, we can discuss moving it to be based on the 11.3 build, which is run on every PR. @seemethere any thoughts about potentially increasing our need for linux16x machines?

The problem is that a large part of distributed is tested via the multigpu tests. If we run these on trunk only, it is very likely that PRs that are green will cause trunk breakage very often.

File: test/distributed/_shard/sharded_tensor/test_sharded_tensor.py [hunk 1]
```diff
@@ -213,6 +213,32 @@ def test_shard_parameter_errors(self):
...
+        weight_og = fc.weight.clone()
+        shard_parameter(fc, 'weight', spec)
+
+        # Verify.
+        self.assertTrue(isinstance(fc.weight, ShardedTensor))
+        local_shards = fc.weight.local_shards()
+        self.assertEqual(1, len(local_shards))
+        self.assertEqual(torch.Size([3, 12]), local_shards[0].tensor.size())
+        self.assertEqual(3, local_shards[0].tensor.size(0))
+        self.assertEqual(12, local_shards[0].tensor.size(1))
```
Comment (rohan-varma) [inline test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:240]: these last 2 lines test the same thing as L238?

File: test/distributed/_shard/sharded_tensor/test_sharded_tensor.py [hunk 2]
```diff
@@ -289,6 +315,51 @@ def test_shard_tensor_errors(self):
         ):
             _shard_tensor(tensor, spec)
 
+    @with_comms(init_rpc=False)
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_shard_tensor_with_meta(self):
```
Comment (rohan-varma) [inline test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:321]: Would we want to add a test showing how the sharded tensor can be initialized?

File: torch/distributed/_shard/api.py [hunk 3]
```diff
@@ -87,6 +90,11 @@ def shard_parameter(
     This method replaces ``module.param_name`` with a
     :class:`torch.distributed._sharded_tensor.ShardedTensor`
 
+    If the parameter is a MetaTensor, then an empty ShardedTensor is
+    materialized as per the provided ``sharding_spec`` and ``src_rank`` is ignored.
+    In this case, it is the user's responsibility to appropriately initialize
```
Comment (rohan-varma) [inline torch/distributed/_shard/api.py:95]: Could the docs add an example of how to do this? Is it with the usual `reset_parameters`?

**Review by rohan-varma — COMMENTED** [review]
Curious, are we planning to support torchdistx's deferred_init APIs as well?

also, would be cool if we can figure out ultimately how to use this in FSDP, i.e. shard the parameter on meta device and only initialize the shards (although its unclear if that will bring about significant perf wins)

File: torch/distributed/_shard/api.py [hunk 4]
```diff
@@ -244,6 +252,11 @@ def shard_module(
 
     Needs to be called on all ranks in an SPMD fashion.
 
+    If the parameter is a MetaTensor, then an empty ShardedTensor is
+    materialized as per the provided ``sharding_spec`` and ``src_rank`` is ignored.
```
Comment (kumpera) [inline torch/distributed/_shard/api.py:256]: We should switch `materialized` with created or initialized.

Materialize is how we call converting a meta tensor into a tensor using a real device.

**Pull request closed without merging**